### DISCUSSION TITLE
@vercel/kv in Next.js not reading environment variables

### DISCUSSION THREAD

**Discussion by nacal** [vercel/community#2281] [Help] [answered]
### Summary

I have created a personal blog site with Next.js. Now I tried to use vercel kv but I get an error that the environment variable cannot be read.

### Additional information

I consulted the documentation and followed these steps Is there any other work that needs to be done besides these?

1. Install the package.
2. Create a KV Database.
3. Connect to a project.
4. Pull environment variables.
5. Use KV store in code.

The error message is as follows
```
@vercel/kv: Missing required environment variables KV_REST_API_URL and KV_REST_API_TOKEN
```
 
I have confirmed that KV_REST_API_URL and KV_REST_API_TOKEN are in .env.development.local. 
It is also tied to the project environment variable, but the same error occurs in the preview environment.

**Comment by amyegan** [comment]
Hi @nacal! When you created the database, did you make it available to all environments including Development?

<img width="561" alt="Database setup form with production, preview, and development selected" src="https://user-images.githubusercontent.com/8247105/236005109-d8f50c13-7d68-4a11-9067-44e192efcd8a.png">

I'm not able to reproduce the error in my test app. I followed the steps you shared, adding a bit of code to get the example data in a serverless function, and ran my project with `vercel dev` without error. If you can create a minimal reproduction, that will help us locate the cause of the missing env var error

**Comment by nacal** [comment, marked as answer]
Hello @amyegan , and thanks for the suggestion.

I was able to identify the likely cause by trying it on a small configuration application.

I have found that this does not occur with server side components.

I want to create something that `kv.incr` by pressing a button, can @vercel/kv achieve this?' When I try to achieve this with a client side component using `'use client'` I get the above error.

**Reply by amyegan** [reply]
Glad you found the cause! You can use `kv.incr` in a component, like the [`ViewCounter`](https://github.com/vercel/examples/blob/main/storage/kv-redis-starter/components/view-counter.tsx) in our [Vercel KV Next.js Starter](https://github.com/vercel/examples/tree/main/storage/kv-redis-starter)

**Reply by nacal** [reply]
Ummm.

I have [Vercel KV Next.js Starter](https://github.com/vercel/examples/tree/main/storage/kv-redis-starter) with One-Click Deploy and have verified that `kv.incr` is working in the Production environment. However, when I pull the repository to the local environment, try `vercel link`, `vercel env pull env.local`, then `pnpm dev` and start up, I get an error. The situation is more complicated and confusing.

**Reply by amyegan** [reply]
Try using `vercel dev` instead of `pnpm dev`. After I shared your situation with the engineering team, they let me know they are working on an issue with KV in SvelteKit local dev https://github.com/vercel/storage/issues/107

I switch between other people's projects a lot, so I tend to use `vercel dev` by default. That seems to be why the project was working for me on local, so using that's my suggested workaround until the KV engineers have a patch ready. 🙇

**Comment by Crateros** [comment]
I ran into this same issue but I understand KV is still in beta - should I only be invoking calls to KV via server side components? Right now I have a trivial example in a client component locally which is causing the same error mentioned by OP.

```
    console.log('saving to KV');
    await kv.set("user_1_session", "session_token_value");
    const session = await kv.get("user_1_session");
    console.log('Session: ', session);
```

`@vercel/kv: Missing required environment variables KV_REST_API_URL and KV_REST_API_TOKEN`

![Screenshot_2023-05-11_at_10_00_04_AM](https://github.com/vercel/community/assets/17658464/6b12d17f-437b-4778-9d83-056d9bc816db)

I have downloaded the appropriate env file from vercel and performed a `vercel link` on the project in question.
I attempted to test KV integration via `vercel dev` instead of `npm run dev` but the same error persisted.

**Reply by amyegan** [reply]
Yes. The environment variables are not intended for client-side use, so the KV store calls should be made in things like server side components or serverless functions.
